source/module_hamilt/ks_lcao/veff_lcao.cpp

```diff
@@ -0,0 +1,162 @@
+}
+
+template<>
+void Veff<OperatorLCAO<double>>::contributeHR()
+{
+
+}
+
+template<>
+void Veff<OperatorLCAO<std::complex<double>>>::contributeHR()
+{
+    ModuleBase::TITLE("Veff", "contributeHR");
+    return;
+}
+
+template<>
+void Veff<OperatorLCAO<double>>::contributeHk(int ik)
+{
+    ModuleBase::TITLE("Veff", "contributeHk");
+    ModuleBase::timer::tick("Veff", "contributeHk");
+    // this->hk_fixed_mock(ik);
+    // this->hk_update_mock(ik);
+
+    //-----------------------------------------
+    //(1) prepare data for this k point.
+    // copy the local potential from array.
+    //-----------------------------------------
+
+    for (int ir = 0; ir < GlobalC::rhopw->nrxx; ir++)
+    {
+        GlobalC::pot.vr_eff1[ir] = GlobalC::pot.vr_eff(GlobalV::CURRENT_SPIN, ir);
+        if(XC_Functional::get_func_type()==3 || XC_Functional::get_func_type()==5)
+        {
+            GlobalC::pot.vofk_eff1[ir] = GlobalC::pot.vofk(GlobalV::CURRENT_SPIN, ir);
+        }
+    }
+
+    //--------------------------------------------
+    // (3) folding matrix,
+    // and diagonalize the H matrix (T+Vl+Vnl).
+    //--------------------------------------------
+
+    if(XC_Functional::get_func_type()==3 || XC_Functional::get_func_type()==5)
+    {
+        Gint_inout inout(GlobalC::pot.vr_eff1, GlobalC::pot.vofk_eff1, this->LM, Gint_Tools::job_type::vlocal_meta);
+        this->GG->cal_vlocal(&inout);
+    }
+    else
+    {
+        Gint_inout inout(GlobalC::pot.vr_eff1, this->LM, Gint_Tools::job_type::vlocal);
+        this->GG->cal_vlocal(&inout);
+    }
+
+    ModuleBase::timer::tick("Veff", "contributeHk");
+}
+
+template<>
+void Veff<OperatorLCAO<std::complex<double>>>::contributeHk(int ik)
+{
+    ModuleBase::TITLE("Veff", "contributeHk");
+    ModuleBase::timer::tick("Veff", "contributeHk");
+    //-----------------------------------------
+    //(1) prepare data for this k point.
+    // copy the local potential from array.
+    //-----------------------------------------
+    for (int ir = 0; ir < GlobalC::rhopw->nrxx; ir++)
+    {
+        GlobalC::pot.vr_eff1[ir] = GlobalC::pot.vr_eff(GlobalV::CURRENT_SPIN, ir);
+        if(XC_Functional::get_func_type()==3 || XC_Functional::get_func_type()==5)
+        {
+            GlobalC::pot.vofk_eff1[ir] = GlobalC::pot.vofk(GlobalV::CURRENT_SPIN, ir);
+        }
+    }
+
+    //--------------------------------------------
+    //(2) check if we need to calculate
+    // pvpR = < phi0 | v(spin) | phiR> for a new spin.
+    //--------------------------------------------
+    if (GlobalV::CURRENT_SPIN == this->GK->get_spin())
+    {
+        // GlobalV::ofs_running << " Same spin, same vlocal integration." << std::endl;
+    }
+    else
+    {
+        // GlobalV::ofs_running << " (spin change)" << std::endl;
+        this->GK->reset_spin(GlobalV::CURRENT_SPIN);
+
+        // if you change the place of the following code,
+        // rememeber to delete the #include
+        if (GlobalV::VL_IN_H)
```

Comment:
I suppose this parameter should work when you are defining the hamiltonian.
That is, do not put vnl in the linked list during initialization